File: infer/tests/codetoanalyze/objc/uninit/arrays.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include <Foundation/Foundation.h>

int define_array_ok() {
  size_t len = 10;
  uint8_t a[len]; // we should not report on array definition

  return 0;
}

int bad1() {
  int a[10];

  return a[2]; // report here as values of the array are not initialized
}

void use_int(int);

int bad2() {
  int a[10];
  use_int(a[2]); // report here as values of the array are not initialized

  return 0;
}

int use_array_ok() {
  int a[10];
  a[1] = 5;

  return a[1];
}

int initialize_array_ok1() {
  int arr[10];
  arr[0] = 5;
}

int initialize_array_ok2() {
  int arr[] = {1, 2};
  arr[0] = 5;
}

void use_array(int[]);

int pass_array_FN() {
  int arr[5];
  use_array(arr); // we do not report here because of intraprocedural analysi a
                  // the array passed by ref
}

int array_length_undef_bad() {
  int x;
  int arr[x];
}

init_array_helper(int*);

void pass_array_address_ok() {
  int x[8];
  init_array_helper(x);
}

File: infer/tests/codetoanalyze/c/uninit/arrays.c
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

int array_good1_FP() {
  int a[10];
  for (int i = 0; i < 10; i++) {
    a[i] = 0;
  }
  return a[2];
}

int array_bad() {
  int a[10];
  for (int i = 0; i < 5; i++) {
    a[i] = 0;
  }
  return a[9];
}

int nondet();

int array_good2_FP() {
  int* a[10];
  int i = 0;
  while (i < 10) {
    if (nondet()) {
      a[i] = 0;
      i++;
    }
  }
  return a[2];
}

int array_good3_FP(int m) {
  int* a[1000];
  int i = 0;
  while (i < 1000 && i < m) {
    if (nondet()) {
      a[i] = 0;
      m--;
    }
  }
  return i > 0 ? a[i - 1] : 42;
}

void mayinit_vpp(const void**);
void use_vp(void*);

void* call_to_mayinit_and_return_good() {
  void* obj[10];
  mayinit_vpp((const void**)obj);
  return obj[0];
}

void call_to_mayinit_and_call_use_good() {
  void* obj[10];
  mayinit_vpp((const void**)obj);
  use_vp(obj[0]);
}
